mlflow/server/handlers.py

```diff
@@ -1103,6 +1112,117 @@ def _default_history_bulk_impl():
+                    f"{MAX_RUNS_GET_METRIC_HISTORY_BULK} run_ids. Received {len(x)} run_ids.",
+                ),
+            ],
+            "metric_key": [_assert_required, _assert_string],
+            "start_step": [_assert_intlike],
+            "end_step": [_assert_intlike],
+            "max_results": [
+                _assert_intlike,
+                lambda x: _assert_intlike_within_range(
+                    int(x),
+                    1,
+                    MAX_RESULTS_PER_RUN,
+                    message=f"max_results must be between 1 and {MAX_RESULTS_PER_RUN}.",
+                ),
+            ],
+        },
+    )
+
+    args = request.args
+    run_ids = request_message.run_ids
+    metric_key = request_message.metric_key
+    max_results = int(args.get("max_results", MAX_RESULTS_PER_RUN))
+
+    store = _get_tracking_store()
+
+    def _get_max_step_for_metric(run_id, metric_key):
+        if hasattr(store, "get_max_step_for_metric"):
+            return store.get_max_step_for_metric(run_id=run_id, metric_key=metric_key)
+        steps = [m.step for m in store.get_metric_history(run_id, metric_key)]
+        return max(steps) if steps else 0
+
+    def _get_sampled_steps(run_ids, metric_key, max_results):
+        # can not fetch from request_message as the default value is 0
+        start_step = args.get("start_step")
+        end_step = args.get("end_step")
+        if start_step is None and end_step is None:
+            max_steps_per_run = [_get_max_step_for_metric(run_id, metric_key) for run_id in run_ids]
+            max_steps = max(max_steps_per_run)
+            # Get sampled steps and append the max step for each run.
+            sampled_steps = _get_sampled_steps_from_steps(0, max_steps, max_results)
+            return list(set(sampled_steps).union(max_steps_per_run))
+        elif start_step is not None and end_step is not None:
+            start_step = int(start_step)
+            end_step = int(end_step)
+            if start_step > end_step:
+                raise MlflowException.invalid_parameter_value(
+                    "end_step must be greater than start_step. "
+                    f"Found start_step={start_step} and end_step={end_step}."
+                )
+            sampled_steps = _get_sampled_steps_from_steps(start_step, end_step, max_results)
+            return list(set(sampled_steps).union([end_step]))
+        else:
+            raise MlflowException.invalid_parameter_value(
+                "If either start step or end step are specified, both must be specified."
+            )
+
+    def _default_history_bulk_interval_impl():
+        steps = _get_sampled_steps(run_ids, metric_key, max_results)
+        metrics_with_run_ids = []
+        for run_id in run_ids:
+            metrics_with_run_ids.extend(
+                store.get_metric_history_bulk_interval_from_steps(
+                    run_id=run_id,
+                    metric_key=metric_key,
+                    steps=steps,
```

Comment:
if the step is empty, what happens? We hit an issue in managed MLflow due to empty steps, so we added extra handling on the situation where the `steps` is an empty list.

Reply:
If `steps` is an empty list (which it actually won't be, because the returned value of `_get_sampled_steps` always unions max_steps_per_run or end_run), then `get_metric_history_bulk_interval_from_steps` will return an empty list as well.